tests/unittests/Foundation/NSOperationTests.mm

```diff
@@ -59,3 +62,264 @@
+   [operation removeObserver:observer forKeyPath:@"isCancelled" context:NULL];
+   [operation removeObserver:observer forKeyPath:@"isExecuting" context:NULL];
+   [operation removeObserver:observer forKeyPath:@"isFinished" context:NULL];
+}
+
+// Test asynchronous subclass for NSOperation
+@interface MyConcurrentOperation : NSOperation
+
+@property (assign, getter = isExecuting) BOOL executing;
+@property (assign, getter = isFinished, readonly) BOOL finished;
+
+@end
+
+@implementation MyConcurrentOperation
+
+@synthesize executing = _executing;
+@synthesize finished = _finished;
+
+- (void)setExecuting:(BOOL)executing
+{
+    [self willChangeValueForKey:@"isExecuting"];
+    [self willChangeValueForKey:@"isFinished"];
+
+    _executing = executing;
+    _finished = !executing;
+
+    [self didChangeValueForKey:@"isExecuting"];
+    [self didChangeValueForKey:@"isFinished"];
+}
+
+- (BOOL)isExecuting
+{
+    return _executing;
+}
+
+- (BOOL)isFinished
+{
+    return _finished;
+}
+
+- (void)start
+{
+    if (self.isCancelled) {
+        return;
+    }
+
+    self.executing = YES;
+    [self doSomething];
+}
+
+- (void)doSomething
+{
+    // Do some async task.
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
+
+        // Do another async task
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
+            // Do some async task.
+            self.executing = NO;
+        });
+    });
+}
+
+@end
+
+TEST(NSOperation, NSOperationConcurrentSubclass) {
+    NSOperationQueue* queue = [[NSOperationQueue new] autorelease];
+
+    NSOperation* operation = [MyConcurrentOperation new];
+
+    [operation setCompletionBlock:^{
+        [operation waitUntilFinished]; // Should not deadlock, but we cannot test this
+        ASSERT_TRUE([operation isFinished]);
+    }];
+
+    [queue addOperation:operation];
+
+    [operation waitUntilFinished];
+
+    ASSERT_TRUE([operation isFinished]);
+    ASSERT_FALSE([operation isExecuting]);
+    ASSERT_NO_THROW([operation release]);
+}
+
+// Test synchronous subclass for NSOperation
+@interface MyNonconcurrentOperation : NSOperation
+@property BOOL didWork;
+@end
+
+@implementation MyNonconcurrentOperation
+
+- (void)main
+{
+    if (self.isCancelled) {
+        return;
+    }
+
+    _didWork = YES;
+}
+
+@end
+
+TEST(NSOperation, NSOperationNonconcurrentSubclass) {
+    NSOperationQueue* queue = [[NSOperationQueue new] autorelease];
+
+    MyNonconcurrentOperation* operation = [MyNonconcurrentOperation new];
+
+    [operation setCompletionBlock:^{
+        [operation waitUntilFinished]; // Should not deadlock, but we cannot test this
+        ASSERT_TRUE([operation isFinished]);
+    }];
+
+    [queue addOperation:operation];
+    [operation waitUntilFinished];
+
+    ASSERT_TRUE([operation isFinished]);
+    ASSERT_FALSE([operation isExecuting]);
+    ASSERT_TRUE([operation didWork]);
```

Comment:
didWork needs to be reset after every successful test else the following test check is not really valid